compiler/src/iree/compiler/Dialect/Stream/IR/StreamOpFolders.cpp

```diff
@@ -507,10 +507,159 @@ void ResourceAllocOp::getCanonicalizationPatterns(RewritePatternSet &results,
 // stream.resource.alloca
 //===----------------------------------------------------------------------===//
 
+namespace {
+
+// Elides transient allocations that have no uses of their resource.
+// This sometimes arises when operations that were using the resource are
+// DCEd by other patterns or passes. The ElideAllocaDeallocaOp pattern will be
+// used after deallocations have been inserted but prior to that point this
+// pattern allows for more eager removal of unused allocations.
+struct ElideUnusedAllocaOp : public OpRewritePattern<ResourceAllocaOp> {
```

Comment:
I wonder if there is a MemEffect that gives this behavior by default. I am assuming you are not making the op have no-side effects cause you dont want the op to be moved around